### DISCUSSION TITLE
FetchError: invalid json response body at...reason: Unexpected token < in JSON at position 0

### DISCUSSION THREAD

**Discussion by davincecode** [vercel/next.js#37454] [Help] [answered]
Hello, Please don't hate me for posting this. I just can't find an answer in each discussion. 

So, I just downloaded this cms-WordPress template to try out Nextjs with WordPress. I also have my WordPress with GraphQL setup on my hosting.

Is there any solution for this error?
FetchError: invalid json response body at http://mydomain.com/graphql reason: Unexpected token < in JSON at position 0

I tried fixing it by setting it up locally.
- Got a few errors but managed to make it work locally. ✅

But then, when I update the env. local back to WORDPRESS_API_URL=http://mydomain.com/graphql, I still get the same error.

Are there any tips or tricks to tackle this? I would appreciate any help. Thank you!

**Comment by icyJoseph** [comment, marked as answer]
> FetchError: invalid json response body at http://mydomain.com/graphql reason: Unexpected token < in JSON at position 0

This means that the fetch call did not receive JSON, but, most likely HTML. That also means that something is wrong in that communication.

Either:
- you are calling it incorrectly, by using an invalid domain, 
- or querying with an invalid schema (though this usually resolves to JSON anyway),
- the `graphql` resolvers are encountering a fatal error, and a 500 is returned with the server custom answer (an HTML page)
- or the endpoint where the `graphql` server runs is not configured correctly.

How, exactly, did you make it work locally? Without a reproduction repository it is a bit hard to tell where things are going wrong.

I think in WP you can find the GraphQL plugin, and there they have an IDE to interact with your graphql server, https://www.wpgraphql.com/docs/interacting-with-wpgraphql/, can you reach it from there?

**Reply by donnydark0** [reply]
I have the same issue. I followed the instructions to install the cms-wordpress step by step. My .env.local looks like this:
WORDPRESS_API_URL="https://sub.domain.com/graphql"
(anonymised sub.domain name)
Then I run npm run dev and get this error after opening localhost:3000
Do I need to do something in Graphql on Wordpress? The instruction just says install the plugin on Wordpress and nothing else.

**Reply by davincecode** [reply]
Thank you, @icyJoseph !

 🎯 the endpoint where the graphql server runs is not configured correctly. 

Also, I followed this from your link and got it to work.

"WPGraphQL adds an endpoint to your WordPress site which can be queried by sending an HTTP Post request specifying at minimum a query as part of the request. The endpoint is either /graphql or /index.php?graphql depending on your WordPress permalink settings."

- [ ]  I updated my `WordPress permalink settings` to `Post name` both offline and on my subdomain and it worked for me.
- [ ]  Made sure the endpoint is working eg. sub.domain.com/graphql

Hope this helps anyone who had the same problem.

**Reply by donnydark0** [reply]
Thanks for the reply. So all you did is change the permalink settings in WP? No changes in GraphQL settings?

That doesn't do anything for me. The WPGraphQL IDE does give correct outputs. I can also reach the sub.domain.com/graphql. But I keep getting the error after compiling.

**Reply by davincecode** [reply]
Would you mind sharing the error you're getting? Maybe I can show you a solution if we're getting the same error.

**Reply by donnydark0** [reply]
Got it working. For some reason it works when I only add https://sub.domain.com without /graphql...

**Reply by BeachBumSpencer** [reply]
> Got it working. For some reason it works when I only add https://sub.domain.com without /graphql...

This then creates an invalid src prop on next/image directory which was resolved by configuring my hostname images within next.config.js, with "_example_.wp.com" the hostname provided by the thrown error, not sure if is the hostname is the same across the board, thus the "example".

From there this issue was solved. Thanks all!

**Reply by saurabhranjan007** [reply]
How would you resolve this in case of a normal REST call?? 

Sharing the codebase and error logs: 

`export async function getServerSideProps() {
  const res = await fetch('https://rickandmortyapi.com/')
  const data = await res.json(JSON.stringify(res))

  return { props: { data } }
}`

`error - FetchError: invalid json response body at https://rickandmortyapi.com/ reason: Unexpected token < in JSON at position 0
    at C:\pproject-files\next-cms-react-apps\wubba-lubba\new-app\node_modules\next\dist\compiled\node-fetch\index.js:1:51227
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async getServerSideProps (webpack-internal:///./pages/index.js:251:18)
    at async Object.renderToHTML (C:\pproject-files\next-cms-react-apps\wubba-lubba\new-app\node_modules\next\dist\server\render.js:490:20)
    at async doRender (C:\pproject-files\next-cms-react-apps\wubba-lubba\new-app\node_modules\next\dist\server\base-server.js:908:38)
    at async cacheEntry.responseCache.get.isManualRevalidate.isManualRevalidate (C:\pproject-files\next-cms-react-apps\wubba-lubba\new-app\node_modules\next\dist\server\base-server.js:1013:28)
    at async C:\pproject-files\next-cms-react-apps\wubba-lubba\new-app\node_modules\next\dist\server\response-cache.js:69:36 {
  type: 'invalid-json',
  page: '/'
}`

**Comment by donnydark0** [comment]
How did you solve it?

**Reply by GlennOu66304** [reply]
Hey, if you already built a serverless function in the development environment area, such as /api/xx, this under the local test is fine. but when you want to deploy it online, then you need to put your serverless api on vercel. 
So the concept is like that the api you build in the local, if you want it to be accessed on production environment, you need to deploy this section online as well.

for details, you can check this article 
[How can I use files in Serverless Functions on Vercel?](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions)  
[Serverless functions with Vercel](https://dev.to/diegoguevara/serverless-functions-with-vercel-287n)
